Fix nix-get-completions, add to nix-repl-mode

rossabaker opened the pull request:
A mashup of python's completion-at-point function and @purcell's [sketch](https://github.com/NixOS/nix-mode/pull/39#issuecomment-360968725).  

This is working for me via `company-capf` on nix-2.3.7.

rossabaker committed:
Fix nix-get-completions, add to nix-repl-mode

rossabaker commented on nix-repl.el:
```diff
@@ -58,33 +68,97 @@
+              (when (nix--accept-process-output
+                     process nix-repl-completion-output-timeout
+                     comint-redirect-finished-regexp)
+                (beginning-of-line)
+                (if (eq (char-after) ?\r)
+                    (cdr
+                     (split-string
+                      (buffer-substring-no-properties
+                       (line-beginning-position) (point-min))
+                      "[ \f\t\n\r\v]+" t))
+                  (search-forward "" nil t)
+                  (backward-char)
+                  (if (eq (char-before) ?\a)
+                      nil
+                    (list (buffer-substring-no-properties (line-beginning-position) (point))))))))
```
This here is the bulk of the original work.

We don't get nice dummy completions like [Python](https://github.com/emacs-mirror/emacs/blob/669b46e6a39bb5ba5d2ed14baebd585af6130ec9/lisp/progmodes/python.el#L3443-L3445). Scala's `sbt` has a nifty `completions` command that sbt-mode can use, and am thinking about proposing similar for `nix repl`.  I'm particularly scared of the fragility of depending on output like `?\a` and what other platforms or .inputrc files might do to this "protocol".

rossabaker commented on nix-repl.el:
```diff
@@ -58,33 +68,97 @@
+  "Completion at point function for Nix using \"nix-repl\".
+See `completion-at-point-functions'."
+  (save-excursion
+    (let ((prefix (and (derived-mode-p 'nix-repl-mode)
```
This is from @purcell's draft, but I am not yet ambitious enough to do this in `nix-mode`.  We might try refreshing the current buffer into a repl, but that could be very slow, and I'm not sure what to do with nix files that require arguments. I think that's more the domain of a [Nix LSP](https://github.com/nix-community/rnix-lsp).

rossabaker reviewed (commented):
Most of my emacs-lisp is trivial and personal. I'm open to any and all criticism. Thanks for looking!

purcell commented:
Nice!

matthewbauer commented on nix-repl.el:
```diff
@@ -21,10 +21,20 @@
-  (setq-local comint-prompt-read-only t))
+  (setq-local comint-prompt-read-only t)
+  (add-hook 'completion-at-point-functions
+            #'nix-repl-completion-at-point nil 'local))
```
This adds it just for the repl buffer right? It should be fine in the Nix buffer too, but we need to make sure it fails gracefully if Nix Repl doesn't exist / isn't running.

rossabaker replied:
Should be only for nix-repl-mode buffers, which I'd expect to only be created by `nix-repl`.  I've tried to make it a bit more graceful in #107.

matthewbauer committed:
Add completion-at-point to nix-repl-mode-map

This makes it a little easier to get to the completion. Taken from the
setting in shell-mode-map.

matthewbauer commented:
Looks good to me, and confirmed it works with Nix 3.0 too!

Nix does have a NIX_GET_COMPLETIONS variable which can get completions of buildable things (try `NIX_GET_COMPLETIONS=2 nix eval nixpkgs#` in Nix 3.0). You could actually use it in the repl, but I don't think it's as thorough as the REPL completion.
